plugins/outputs/application_insights/application_insights.go

```diff
@@ -0,0 +1,422 @@
+## and the ai.cloud.roleInstance context tag to the value of kubernetes_pod_name property.
+## For list of all context tag keys see https://github.com/Microsoft/ApplicationInsights-Go/blob/master/appinsights/contracts/contexttagkeys.go
+`
+	is32Bit        bool
+	is32BitChecked bool
+)
+
+func (a *ApplicationInsights) SampleConfig() string {
+	return sampleConfig
+}
+
+func (a *ApplicationInsights) Description() string {
+	return "Send telegraf metrics to Azure Application Insights"
+}
+
+func (a *ApplicationInsights) Connect() error {
+	if a.transmitter == nil && a.InstrumentationKey != "" {
+		a.transmitter = NewAppinsightsTransmitter(a.InstrumentationKey)
+	}
+
+	if a.EnableDiagnosticLogging && a.diagMsgSubscriber != nil {
+		a.diagMsgListener = a.diagMsgSubscriber.Subscribe(func(msg string) error {
+			logOutputMsg("%s", msg)
+			return nil
+		})
+	}
+
+	return nil
+}
+
+func (a *ApplicationInsights) Write(metrics []telegraf.Metric) error {
+	if len(metrics) == 0 || a.transmitter == nil || a.InstrumentationKey == "" {
+		return nil
+	}
+
+	for _, metric := range metrics {
+		allMetricTelemetry := a.createTelemetry(metric)
+		for _, telemetry := range allMetricTelemetry {
+			a.transmitter.Track(telemetry)
+		}
+	}
+
+	return nil
+}
+
+func (a *ApplicationInsights) Close() error {
+	if a.diagMsgListener != nil {
+		// We want to listen to diagnostic messages during closing
+		// That is why we stop listening only after Close() ends (or a timeout occurs)
+		defer a.diagMsgListener.Remove()
+	}
+
+	if a.transmitter == nil {
+		return nil
+	}
+
+	select {
+	case <-a.transmitter.Close():
+		logOutputMsg("Closed")
+	case <-time.After(a.Timeout.Duration):
+		logOutputMsg("Close operation timed out after %v", a.Timeout.Duration)
+	}
+
+	return nil
+}
+
+func init() {
+	outputs.Add("application_insights", func() telegraf.Output {
+		return &ApplicationInsights{
+			Timeout:           internal.Duration{Duration: time.Second * 5},
+			diagMsgSubscriber: appinsightsDiagnosticsMessageSubscriber{},
+			// It is very common to set Cloud.RoleName and Cloud.RoleInstance context properties, hence initial capacity of two
+			ContextTagSources: make(map[string]string, 2),
+		}
+	})
+}
+
+func canBeTreatedAsAggregate(metric telegraf.Metric) bool {
+	// If we can identify the Sum and Count properties, then we are good
+	hasSum := metric.HasField("value") || metric.HasField("sum")
+	hasCount := metric.HasField("count") || metric.HasField("samples")
+	return hasSum && hasCount
+}
+
+func (a *ApplicationInsights) createTelemetry(metric telegraf.Metric) []appinsights.Telemetry {
+	if canBeTreatedAsAggregate(metric) {
+		aggregateTelemetry, usedFields := a.createAggregateMetricTelemetry(metric)
+		if aggregateTelemetry != nil {
+			remainingTelemetry := a.createTelemetryForUnusedFields(metric, usedFields)
+			retval := make([]appinsights.Telemetry, 1, 1+len(remainingTelemetry))
+			retval[0] = aggregateTelemetry
+			retval = append(retval, remainingTelemetry...)
```

Comment:
Does the aggregate telemetry need to be first?  If not just append to remainingTelemetry and name variable `telemetry`.